src/loop_logic.rs

```diff
@@ -56,6 +58,7 @@ pub struct RegistrationToken {
 pub(crate) struct LoopInner<'l, Data> {
     pub(crate) poll: RefCell<Poll>,
     pub(crate) sources: RefCell<Slab<Rc<dyn EventDispatcher<Data> + 'l>>>,
+    pub(crate) sources_with_additional_lifetime_events: AdditionalLifetimeEventsSet,
```

Comment:
Could we just duplicate the way sources are defined instead of rolling new abstraction? And create an iterator which simply chains `both` when needed?